### PR TITLE
Make it possible for p/g pins to place at all sides

### PR DESCRIPTION
**Pull request by TimeExceed**
Besides that,

1. circuits connecting core rings and p/g pins go a little bit deeper into p/g pins to improve connectivity.
2. p/g pins are connected to core rings at all layers.

Signed-off-by: TimeExceed <tyf00@aliyun.com>

### CODE REVIEW

**Comment by QuantamHD** [comment]
@TimeExceed Unfortunately I'm just a community member so I can't sign off on your PR. @Colin-Holehouse do you know when you would be able to give this a once over?

**Comment by maliberty** [comment]
@Colin-Holehouse please review

**Comment by maliberty** [comment]
@arlpetergadfort anything to incorporate here for your version?

**Comment by arlpetergadfort** [comment]
> @arlpetergadfort anything to incorporate here for your version?

I think this is covered by `-connect_to_pads` in the `add_pdn _ring` command, it will ensure all the iterms facing the core are connected (if possible and connected to the ring nets in the database) to the ring on all layers with connect assignments.

File: src/pdn/test/four_sides/pad.lef [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+        RECT 27.155 75.635 27.845 80 ;
+        RECT 29.755 75.635 30.445 80 ;
+        RECT 32.355 75.635 33.045 80 ;
+        RECT 34.955 75.635 35.645 80 ;
+        RECT 37.555 75.635 38.245 80 ;
+        RECT 40.155 75.635 40.845 80 ;
+        RECT 42.755 75.635 43.445 80 ;
+        RECT 45.355 75.635 46.045 80 ;
+        RECT 47.955 75.635 48.645 80 ;
+        RECT 3.28 77.9 50.735 80 ;
```
Comment (Colin-Holehouse) [inline src/pdn/test/four_sides/pad.lef:108]: This pin shape overlaps all the previously defined pin shapes in the same layer. Seems like the smaller pin shapes should be removed.

File: src/pdn/test/four_sides/pdn.cfg [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    gnd_pads {
+        VSS {GPAD_LONG GPAD_SHORT}
+    }
+    pwr_pads {
+        VDD {PPAD_LONG PPAD_SHORT}
+    }
+    connect {
+        {metal1 metal2} {metal2 metal3} {metal3 metal4}
+    }
+}
```
Comment (Colin-Holehouse) [inline src/pdn/test/four_sides/pdn.cfg:35]: Its a strange design that has no rails or straps. The straps do interact with the connections to the core rings, so whouls probably be included.

File: src/pdn/src/PdnGen.tcl [hunk 3]
```diff
@@ -3230,36 +3273,6 @@ proc generate_via_stacks {l1 l2 tag connection} {
...
-        continue
-      }
-    } else {
-      if {$width < [get_grid_wire_width $layer2]} {
-        # If the intersection doesnt cover the whole width of the top level wire, then ignore
-        utl::warn "PDN" 43 "No via added at ([expr 1.0 * $xMin / $def_units] [expr 1.0 * $yMin / $def_units] [expr 1.0 * $xMax / $def_units] [expr 1.0 * $yMax / $def_units]) because the full width of $layer2 ([expr 1.0 * [get_grid_wire_width $layer2] / $def_units]) is not covered by the overlap."
-        continue
-      }
-    }
-
```
Comment (Colin-Holehouse) [inline src/pdn/src/PdnGen.tcl:3262]: This code specifically checks that in order to connect two layers together, that there is a full overlap of the lines. A parial overlap can result in problems with via sizing, and are likely the result of a poor floorplan, or interaction of macros cutting channels into the design.
I believe this code needs to be retained.

File: src/pdn/src/PdnGen.tcl [hunk 4]
```diff
@@ -399,6 +399,47 @@ variable voltage_domains {
...
+  foreach layer $layers {
+    set max_width [$layer getMaxWidth]
+    if {$max_width < 2147483647} {
+      return $max_width
+    }
+  }
+
+  variable default_maxwidth
+  return $default_maxwidth
+}
```
Comment (Colin-Holehouse) [inline src/pdn/src/PdnGen.tcl:431]: If there is no MAXWIDTH rule defined for the layer, then the width of the pin should be used.

**Review by Colin-Holehouse — CHANGES_REQUESTED** [review]
It's not very clear what this change is trying to achieve.

The testcase seems somewhat contrived, consisting of just pads and stdcell rows, and the LEF descriptions provided seem to define overlapping pins, which is likely to be problematic.

Some unrelated changes have been done, to cope with the fact that the testcase has no straps defined (unrealistic setup, but the changes are benign), and some code to check layer overlap sizes has been removed that I believe is a core part of how pdn works.

Best I can make out, the code is primarily looking to cope with wide p/g pins on the pads as they connect to the core rings, and tries to automate the addition of a number of max width straps from the pad pin to the core ring when the pin on the pad is very wide.

Its not clear to me that the suggested approach here is always the one that is desired, but there are no alternative approaches allowed (e.g. the current approach could still be appropriate).

**Comment by TimeExceed** [comment]
> It's not very clear what this change is trying to achieve.
> 
> The testcase seems somewhat contrived, consisting of just pads and stdcell rows, and the LEF descriptions provided seem to define overlapping pins, which is likely to be problematic.
> 
> Some unrelated changes have been done, to cope with the fact that the testcase has no straps defined (unrealistic setup, but the changes are benign), and some code to check layer overlap sizes has been removed that I believe is a core part of how pdn works.
> 
> Best I can make out, the code is primarily looking to cope with wide p/g pins on the pads as they connect to the core rings, and tries to automate the addition of a number of max width straps from the pad pin to the core ring when the pin on the pad is very wide.
> 
> Its not clear to me that the suggested approach here is always the one that is desired, but there are no alternative approaches allowed (e.g. the current approach could still be appropriate).

You are right. It is not very clear. I shall split this pr into several ones.

Thank u very much.

**Pull request closed without merging**